### DISCUSSION TITLE
Projection with huge dataset

### DISCUSSION THREAD

**Discussion by dilab** [spatie/laravel-event-sourcing#432] [Q&A]
Our event stored(`stored_events` table) has grown to `12861457` records. which is in 9.9 GiB in size.

We are facing a challenge when generating some reports. 

It will need to read all the events of an aggregate and build the reports (this is correct way of doing it). However due to the size of the data, it has become slow, and it also jams the MySQL, because I have a report server (which spins up 10 workers) to speed up the report generation.

My Question is what I can do to optimize it?

1. Setup a separate MySQL database replicate ?
2. Using a different database engine which is better at I/O?

**Comment by etahamer** [comment]
@dilab Have you thought about using projectors and projections to aggregate data into more convenient form for your report generation process?

**Reply by dilab** [reply]
definitely thought about it. However our reports require to fetch point of time data.

**Reply by sjardim** [reply]
The event sourcing concept you may look for is a snapshot. It's a way to "split" the events into chunks based on the business's requirements. 

See this great talk giving a nice example: https://www.youtube.com/watch?v=gG6DGmYKk4I

**Reply by dilab** [reply]
when building a reports(point of time data), our reports need to examine all the past events.

**Reply by sjardim** [reply]
I understood that. Snapshots are a way of setting "points in time" for the events and their "results." When creating each snapshot, you can calculate all the metrics you want to report on at that period. Each snapshot will then "feed" into the next, so you have all the past events data in the end. What you are doing is just pre-calculating them in chunks to save time. You don't need to recalculate them again for all the reports thereafter. 

It's difficult to give examples without knowing your domain, so that's why I sent that video as an example.

**Reply by dilab** [reply]
Hi @sjardim thanks for the feedback.

Unfortunately, the business requires to dynamically set the query time. meaning the user will select which period of the report to generate on the fly.

**Comment by sjardim** [comment]
Ah, okay. So that presents a problem indeed. Well, when you figure out a solution, please care to post it here if possible so we can all learn from it! :) Best of luck!

**Reply by dilab** [reply]
our current solution is to have a separate database for the reports table, so it won't interfere the normal projection table.

**Comment by sjardim** [comment]
A separate DB is a common solution for this scenario and a good one. The beauty of the projections is that you can create tables (etc.) the way you need for the reports and create better indexes to handle the query load. And as you having performance issues on MySQL, check if the indexes are really being used (it can be tricky).

**Reply by dilab** [reply]
yea.. except that it takes hours to build a report.

**Comment by Gummibeer** [comment]
7.732.919 events with 85.13 GB and zero problems with performance here. We are recording with requests, jobs and webhooks (resulting in 2-3 events per second in average) without snapshots and also rebuilding the same aggregate several times per runtime because of projectors/reactors and everything is fine, only issue we hit was the [MySQL sort buffer](https://github.com/spatie/laravel-event-sourcing/issues/515) which is solved now.

So it's not a package issue or of the concept in general but seems like your DB server could need an upgrade. The only case we are seeing performance issues is with aggregates holding over 300 events - but only when most of them are big ones with big event properties as the retrieving and aggregating of all these rows multiple times per runtime takes time and memory and bandwith.

But as you have ~1.5 times the event count but only ~12% the storage size the bandwith and memory isn't really your issue. Have you analyzed your queries? Like what is really slowing your DB? 13M rows is absolute rookie numbers for MySQL - so it's likely the queries being not optimized.
